clang/lib/Sema/SemaDecl.cpp

```diff
@@ -16555,7 +16555,7 @@ Decl *Sema::ActOnFinishFunctionBody(Decl *dcl, Stmt *Body,
     checkTypeSupport(FD->getType(), FD->getLocation(), FD);
 
   // Handle free functions.
-  if (LangOpts.SYCLIsDevice && FD->hasAttr<SYCLDeviceAttr>() && Body &&
+  if (FD && LangOpts.SYCLIsDevice && FD->hasAttr<SYCLDeviceAttr>() && Body &&
```

Comment:
Suggestion:

Add:
``  if (!FD)
      return dcl;
``
before 16545, and remove the FD checks on lines 16545, 16554 and the new 16558.

Reply:
Thanks @premanandrao for the reviews and suggestion.

Done